spark/BidResponseJob.scala

```diff
@@ -0,0 +1,300 @@
+  {
+    def addPrefixGroup(extractor: (T)=>(Option[Int],Option[String]), noAll: Boolean=false, noAllRtb: Boolean=false): TypedPipe[(PrefixGroup,T)] = {
+      val unacceptableBidders = (noAll, noAllRtb) match {
+        case (true,true) => List("ALL","ALL-RTB")
+        case (false,true) => List("ALL-RTB")
+        case (true,false) => List("ALL")
+        case (false,false) => Nil
+      }
+      pipe.flatMap( t => (extractor(t) match {
+        case (None, None) => Seq(("ALL", "NO-HASH"))
+        case (None, Some(hash))  => Seq(("ALL", mappingSelector.getGroupPrefix(hash, MappingSelector.CommonGroupings).getOrElse("OUTSIDE-GROUPS")))
+        case (Some(bidder), None) if mappingSelector.excludedFromWatermarking.contains(bidder.toString()) =>
+          Seq((bidder.toString(), if (mappingSelector.groupingNames.contains(bidder.toString())) "NO-HASH" else "UNKNOWN-BIDDER"))
+        case (Some(bidder), None) => Seq(("ALL", "NO-HASH"),
+          ("ALL-RTB", "NO-HASH"),
+          (bidder.toString(), if (mappingSelector.groupingNames.contains(bidder.toString())) "NO-HASH" else "UNKNOWN-BIDDER"))
+        case (Some(bidder), Some(hash)) if mappingSelector.excludedFromWatermarking.contains(bidder.toString()) =>
+          Seq((bidder.toString(), mappingSelector.getGroupPrefix(hash, bidder.toString()).getOrElse("UNKNOWN-BIDDER")))
+        case (Some(bidder), Some(hash)) => Seq(("ALL", mappingSelector.getGroupPrefix(hash, MappingSelector.CommonGroupings).getOrElse("OUTSIDE-GROUPS")),
+          ("ALL-RTB", mappingSelector.getGroupPrefix(hash, MappingSelector.CommonGroupings).getOrElse("OUTSIDE-GROUPS")),
+          (bidder.toString(), mappingSelector.getGroupPrefix(hash, bidder.toString()).getOrElse("UNKNOWN-BIDDER")))
+      }).filterNot( e => unacceptableBidders.contains(e._1) ).flatMap( pg => Seq((PrefixGroup(pg._1, pg._2), t), (PrefixGroup(pg._1, "ALL"), t)) ) )
+    }
+  }
+
+  import com.twitter.algebird._
+
+  private implicit def tuple3Semigroup[A:Semigroup,B:Semigroup,C:Semigroup]: Semigroup[(A,B,C)] = new Semigroup[(A,B,C)] {
+    def plus(l: (A,B,C), r: (A,B,C)): (A,B,C) = (implicitly[Semigroup[A]].plus(l._1, r._1),
+      implicitly[Semigroup[B]].plus(l._2, r._2),
+      implicitly[Semigroup[C]].plus(l._3, r._3))
+  }
+
+  private val smAggregator = SketchMap.aggregator[String, Long](SketchMapParams[String](1, 0.001, 1E-10, 50){ s => s.toArray.map( _.toByte ) })
+
+  private implicit val smSemigroup: Semigroup[SketchMap[String,Long]] = new Semigroup[SketchMap[String,Long]] {
+    def plus(l: SketchMap[String,Long], r: SketchMap[String,Long]) = smAggregator.reduce(l, r)
+  }
+
+  private implicit class MapAdd[X,Y](m: Map[X,Y]) extends Serializable {
+    def +|+(other: Map[X,Y]) = m ++ other.map( e => if (m.contains(e._1)) throw new RuntimeException(s"Found key ${e._1} in both maps") else e )
+  }
+
+  private[BidResponsesJob] implicit class Aggregator[T](val pipe: TypedPipe[(PrefixGroup,T)]) extends Serializable
```

Comment:
Is it on purpose we have TypedPipe within spark code?

Reply:
no... i converted from scalding and missed that. will check the whole thing again